rfcs/react-components/convergence/custom-styling.md

```diff
@@ -0,0 +1,475 @@
+# RFC: Custom Styling
+
+## Contributors
+
+- Geoff Cox
+- Mason Tejera - MADS use cases
+- Ben Howell - easier recomposition
+- Micah Godbolt - component CSS vars
+
+## Summary
+
+Fluent UI React v9 is built today to adopt the Fluent 2 design as its default style. We strongly recommend that
+Microsoft applications leverage the default style and limit customization to a brand-based theme.
+
+Currently, v9 has mechanisms to support several customization scenarios:
+
+- Custom themes => Customize style values for all components
+- `classname` prop => Customize the style of one instance of a component
+- Extended design tokens => Support theme for a custom component or custom class/style
+- Hooks recomposition => Create a new component customizing the behavior, style, or rendering of an existing component.
+
+See the appendix for more detailed analysis of the capabilities and limitations of each of these mechansisms.
+
+## Problem
+
+Partners migrating from v0/v8 to v9 have brought up that v9's custom styling does not provide the same granularity of
+customization as previous versions. See the appendix for detailed partner use cases and scenarios.
+
+Partners producing component libaries or component design surfaces are the most concerned about the lack of
+customization. They would like to customize the style of a single type of component without changing the style of
+other components.
+
+Example: Change the border radius and border width of all Button components without changing those styles for
+Input components.
+
+## Option A: React.Context useCustomStyles hooks
+
+Each component could check context for a useCustomStyle hook within the useComponentStyle hook.
+If this hook was defined, it is called at the end of the method before returning state.
+
+> These example use Component as a substitute for a component like Button or Avatar.
+
+```tsx
+const customStyles = useContextSelector(ComponentCustomStylesContext, context => context?.useCustomStyles);
+
+//...component styling here
+
+return customStyles ? customStyles(state) : state;
```

Comment:
nit: Technically, `customStyles()` would need to be unconditionally called hook to comply with rules of hooks and to allow `makeStyles` to be used in it.

Reply:
Great point! I'll think about this some more while I'm playing with this option.  It certainly puts a wrench in conditionally calling it.

Reply:
It's another question actually: is there need to have `useCustomStyle` hook of it could be just a `customStyle` function? Using hooks will have limitations.

Reply:
I think it has to be a hook if they use Griffel right? makeStyles returns a hook they would call to apply class names to state just like the components useStyle hook.